### ISSUE TITLE
[4][com_finder] Some Search results lead to 404 errors

### ISSUE THREAD

**Issue by PhilETaylor** [joomla/joomla-cms#30968]
### Steps to reproduce the issue

Forked from issue: https://github.com/joomla/joomla-cms/issues/30967

Set up is as per the above issue

Note in the db I have three categories after installing the sample data 

Uncategorised = ID 2
Blog = ID 8
Help = ID 9

Look at the search results ...

<img width="879" alt="Screenshot 2020-10-07 at 01 45 48" src="https://user-images.githubusercontent.com/400092/95275323-1220ca00-0840-11eb-840a-344a3b4649e7.png">

### Expected result

No invalid links to invalid categories in the search results that when clicked give 404 errors

### Actual result

404 errors 

http://127.0.0.1:4444/index.php?view=category&id=7

http://127.0.0.1:4444/index.php?view=category&id=3

<img width="818" alt="Screenshot 2020-10-07 at 01 55 48" src="https://user-images.githubusercontent.com/400092/95275385-40060e80-0840-11eb-8ec2-a4ec9df09107.png">

**Comment by Hackwar** [comment]
This isn't an issue of Smart Search, but of the routing. There simply is no menu route to the mentioned categories and thus it falls back on the home menuitem. That then again results in a 404 URL. Don't know how to fix this yet, but marking it as a com_finder/Smart Search issue is missing the point.

**Comment by PhilETaylor** [comment]
You missed the point ... 

These are the ONLY categories in my site:

Uncategorised = ID 2
Blog = ID 8
Help = ID 9

Yet in the search results there are links to categories with id 7 and 3

All I have done is install Joomla 4 and Sample Blog Data and then Index the site. THE CATEGORIES with id 7 and 3 DO NOT EXIST and therefore these categories should never be in the Index...

**Comment by ghost** [comment]
![Screen Shot 2020-10-10 at 06 48 10](https://user-images.githubusercontent.com/71960554/95646035-a4390480-0ac4-11eb-9293-ce9ae088ec8f.png)

**Comment by infograf768** [comment]
@PhilETaylor 
if you display `administrator/index.php?option=com_finder&view=index` and filter by category, do you find your ids?
Likely Uncategorised for users, banners, or newsfeeds I guess.

**Comment by PhilETaylor** [comment]
I was reading the MySQL table directly 

I'll try again today

**Comment by PhilETaylor** [comment]
@gostn You have obviously not followed the steps to replicate the bug FROM A CLEAN SLATE as the "Blog Sample Data" doesn't install other languages. 

@infograf768 I have checked the database direct again today - and there many more categories than the three I saw the other night :-( 

<img width="744" alt="Screenshot 2020-10-10 at 11 15 14" src="https://user-images.githubusercontent.com/400092/95652558-e6c10800-0ae9-11eb-9faa-66391f0dd4c2.png">


After INDEXING the "Blog Sample Data" right after installing it you get this:

<img width="1346" alt="Screenshot 2020-10-10 at 11 12 15" src="https://user-images.githubusercontent.com/400092/95652515-77e3af00-0ae9-11eb-9476-561074ec9bb7.png">

Then on the frontend (using the bug in https://github.com/joomla/joomla-cms/issues/30967) you see:

<img width="762" alt="Screenshot 2020-10-10 at 11 15 47" src="https://user-images.githubusercontent.com/400092/95652567-f6405100-0ae9-11eb-9859-945327ecd95b.png">

So the BUG here is that the links are missing the `option=com_XXX` param in the URLS.

**Comment by ghost** [comment]
> You have obviously not followed the steps to replicate the bug FROM A CLEAN SLATE as the "Blog Sample Data" doesn't install other languages.

thats why i filter languages to ´all´.

**Comment by PhilETaylor** [comment]
The only way to test something is from a clean slate and then following the "steps to reproduce" provided. 

Working on a site with changes already can cause all kinds of side-issues. Maybe not in this case on this issue, but it will trip you up more times than you will want to admit. Always start from a clean known state :)

**Comment by ghost** [comment]
> Working on a site with changes already

the "changes":

![image](https://user-images.githubusercontent.com/71960554/95652757-cac26400-0af3-11eb-9b41-7b7ebf3f8d18.png)

**Comment by infograf768** [comment]
There is a bug indeed 
> So the BUG here is that the links are missing the option=com_XXX param in the URLS.

Correct. And I also wonder why you get these results when searching for `PhilTaylor`

**Comment by PhilETaylor** [comment]
@gostn Yes but the steps published to reproduce say to use the "Blog Sample Data" and not the "Multilingual Sample Data" - for this issue it made no difference, but in future it might 

@infograf768 Maybe its all related to the parent issue this forked from - maybe I should never see these results if the bug #30967 was fixed?

**Comment by ghost** [comment]
> Yes but the steps published to reproduce say to use the "Blog Sample Data" and not the "Multilingual Sample Data

i used both as always so what.

**Comment by PhilETaylor** [comment]
You can do what ever you want... all I am saying is that the correct way to test is from a known clean state, and that is to follow the "Steps to reproduce"... this time you got away with it, that might not always be the case :) Up to you though.

**Comment by Hackwar** [comment]
@PhilETaylor I did understand you correctly and I followed your instructions. As you noticed yourself, there are more categories in your site than those 2 that you saw. I assumed that it is a problem of the router and not the search. When enabling SEF URLs, the router tries to somehow connect the URL to a menu item. If that fails, it falls back on the home menu item. URL query parameters, which are already given by the menu item, are removed. In your case that is the &option=com_content, which leaves the &view=category&id=7. 

However on deeper inspection, it is an issue of the categories indexing plugin. The plugin was indexing categories for components without a frontend view for them, thus without anything to display. Someone thought it might be a good idea to simply fall back on the com_content RouteHelper. So the given URL was `index.php?option=com_content&view=category&id=7`, while it actually should have been for com_users, which doesn't have a frontend category view and thus this category shouldn't even be indexed at all.

The fix for this is in #31041
